### PR TITLE
Introducing the VOF equation and TwoPhaseTransport model

### PR DESCRIPTION
**Pull request by gdeskos**

### CODE REVIEW

File: amr-wind/equation_systems/vof/vof_ops.H [hunk 1]
```diff
@@ -0,0 +1,89 @@
+#ifndef VOF_OPS_H
+#define VOF_OPS_H
+
+#include "amr-wind/equation_systems/vof/vof.H"
+
+namespace amr_wind {
+namespace pde {
+
+/** Right-hand side (RHS) evaluator for VOF transport equation
+ *  \ingroup vof
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/vof_ops.H:10]: The `vof` group is not defined anywhere.

File: amr-wind/physics/multiphase/MultiPhase.H [hunk 2]
```diff
@@ -14,6 +14,8 @@
  */
 namespace amr_wind {
 
+enum class multiphase_model { VOF, LS, CLSVOF };
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/MultiPhase.H:17]: Use camel case here
```suggestion
enum class MultiphaseModel { VOF, LS, CLSVOF };
```

File: amr-wind/physics/multiphase/MultiPhase.H [hunk 3]
```diff
@@ -56,6 +61,9 @@ private:
 
     // Dynamic Viscosity for Fluid 2
     amrex::Real m_mu2{1.0};
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/MultiPhase.H:63]: Use Transport model here.

File: amr-wind/physics/multiphase/MultiPhase.cpp [hunk 4]
```diff
@@ -10,16 +10,17 @@ MultiPhase::MultiPhase(CFDSim& sim)
...
-    // Register levelset equation
-    auto& levelset_eqn = sim.pde_manager().register_transport_pde("Levelset");
-    // Defer getting levelset field until PDE has been registered
-    m_levelset = &(levelset_eqn.fields().field);
-
     amrex::ParmParse pp_multiphase("MultiPhase");
     pp_multiphase.query("density_fluid1", m_rho1);
     pp_multiphase.query("density_fluid2", m_rho2);
     pp_multiphase.query("viscosity_fluid1", m_mu1);
     pp_multiphase.query("viscosity_fluid2", m_mu2);
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/MultiPhase.cpp:17]: Use transport model here.

File: amr-wind/physics/multiphase/MultiPhase.H (hunk 2 above)
Comment (gdeskos) [reply amr-wind/physics/multiphase/MultiPhase.H]: Revomed for the moment

File: amr-wind/physics/multiphase/MultiPhase.cpp (hunk 4 above)
Comment (gdeskos) [reply amr-wind/physics/multiphase/MultiPhase.cpp]: Two phase transport model is implemented + tested

File: amr-wind/equation_systems/vof/vof_ops.H (hunk 1 above)
Comment (sayerhs) [reply amr-wind/equation_systems/vof/vof_ops.H:10]: This hasn't been fixed yet?

File: amr-wind/physics/multiphase/VolumeFractions_K.H [hunk 5]
```diff
@@ -0,0 +1,231 @@
+#ifndef VolumeFraction_K_H_
+#define VolumeFraction_K_H_
+
+#include <AMReX_FArrayBox.H>
+#include <cmath>
+
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/VolumeFractions_K.H:6]: Should be in `amr_wind` namespace. Don't use `_K.H` in newly introduced files. We don't support 2D. Prefer reserving camelcase names for headers with class definitions.

File: amr-wind/physics/multiphase/VolumeFractions_K.H [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+        mm = m12;
+        V3 = 0.5 * mm / m3;
+    }
+
+    amrex::Real ch = std::min(cc, 1.0 - cc);
+
+    // Original method
+    amrex::Real AL3D, p, q, p12, teta, cs;
+    if (ch < V1) {
+        AL3D = std::pow(pr * ch, 1. / 3.); // case (1)
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/VolumeFractions_K.H:108]:
```suggestion
        AL3D = std::cbrt(pr * ch); // case (1)
```

File: amr-wind/physics/multiphase/VolumeFractions_K.H [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+/** Computes the "CUT VOLUME" V0 given r0, dr0 and
+ *  m_1*x_1 + m_2*x_2 + m_3*x_3 = alpha
+ *  (1) move origin to r0 along r ;
+ * (2) reflect parallelepiped;
+ * (3) limit alpha (0<= al0 <=0.5);
+ * (4) order coefficients: b1<b2<b3;
+ * (5) calculate volume (NOTE: it is assumed:s0=t0=0; ds0=dt0=1.)
+ */
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE amrex::Real FL3D(
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/VolumeFractions_K.H:144]: Use a better name for this function.

File: amr-wind/physics/multiphase/VolumeFractions_K.H [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+          2.0 * (c(i - 1, j, k + 1) + c(i + 1, j, k + 1) + c(i, j - 1, k + 1) +
+                 c(i, j + 1, k + 1)) +
+          4.0 * c(i, j, k + 1);
+    mz = mm1 - mm2;
+}
+/* Computes alpha: m1*x1 + m2* x2 + m3*x3 = alpha
+ * given that m1+m2+m3=1 (m1,m2,m3>0) and the volumetric fraction cc
+ */
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE amrex::Real
+AL3D(amrex::Real mx, amrex::Real my, amrex::Real mz, amrex::Real cc) noexcept
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/VolumeFractions_K.H:58]: use better name for this function.

File: amr-wind/transport_models/TwoPhaseTransport.H [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+                const amrex::Array4<amrex::Real>& phi =
+                    levelset(lev).array(mfi);
+                const amrex::Array4<amrex::Real>& visc = (*mu)(lev).array(mfi);
+                const amrex::Real eps = std::cbrt(2. * dx[0] * dx[1] * dx[2]);
+                const amrex::Real mu1 = m_mu1;
+                const amrex::Real mu2 = m_mu2;
+                amrex::ParallelFor(
+                    vbx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                        amrex::Real H;
```
Comment (sayerhs) [inline amr-wind/transport_models/TwoPhaseTransport.H:73]: use better name for this variable...

File: amr-wind/physics/multiphase/VolumeFractions_K.H (hunk 5 above)
Comment (gdeskos) [reply amr-wind/physics/multiphase/VolumeFractions_K.H]: Done !

File: amr-wind/physics/multiphase/VolumeFractions_K.H (hunk 6 above)
Comment (gdeskos) [reply amr-wind/physics/multiphase/VolumeFractions_K.H]: Done!

**Review by gantech — APPROVED** [review]

**Comment by michaeljbrazell** [comment]
@gdeskos This causes a diff for the abl_ksgsm84_godunov reg test

12: 
12:             variable name            absolute error            relative error
12:                                         (||A - B||)         (||A - B||/||A||)
12:  ----------------------------------------------------------------------------
12:  level = 0
12:  density                                          0                         0
12:  gpx                                9.400480394e-15           6.299509142e-13
12:  gpy                                7.247535905e-15           4.607868348e-13
12:  gpz                                1.276211373e-09           4.465825541e-09
12:  mu_turb                            1.318189935e-08           1.994962226e-08
12:  p                                  2.610966199e-08           4.237782796e-10
12:  temperature                        4.108449048e-08           1.330806453e-10
12:  tke                                5.385566992e-11           5.341979411e-10
12:  velocityx                          1.776356839e-14           2.512834377e-15
12:  velocityy                          1.953992523e-14           3.212183338e-15
12:  velocityz                          8.054668044e-14           1.960186652e-12
12:  velocity_mueff                     1.318189935e-08           1.994932035e-08
1/1 Test #12: abl_ksgsm84_godunov ..............***Failed    5.51 sec

**Comment by sayerhs** [comment]
@michaeljbrazell @gantech @gdeskos 

With #282 the diffs with this PR are down really small values

```
Test project /Users/sanantha/exawind/source/amr-wind/build_base
    Start 12: abl_ksgsm84_godunov
1/1 Test #12: abl_ksgsm84_godunov ..............***Failed   11.29 sec

            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 density                                          0                         0
 gpx                                2.387423592e-15           1.606464101e-13
 gpy                                2.387423592e-15           1.524784549e-13
 gpz                                3.497202528e-15            1.22377035e-14
 mu_turb                            8.026912468e-14            1.21480796e-13
 p                                  3.410605132e-13           5.535680475e-15
 temperature                        2.273736754e-13           7.365075022e-16
 tke                                2.303712776e-15           2.285099292e-14
 velocityx                          5.329070518e-15           7.538873777e-16
 velocityy                          4.440892099e-15           7.300663311e-16
 velocityz                          9.044360313e-15           2.201042395e-13
 velocity_mueff                     8.026912468e-14           1.214789575e-13
```

**Pull request merged**